sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -477,6 +478,25 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = removeRedundantAliases(plan, AttributeSet.empty)
 }
 
+/**
+ * Remove redundant aggregates from a query plan. A redundant aggregate is an aggregate whose
+ * only goal is to keep distinct values, while its parent aggregate would ignore duplicate values.
+ */
+object RemoveRedundantAggregates extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case upper @ Aggregate(_, _, lower: Aggregate) if lowerIsRedundant(upper, lower) =>
+      upper.copy(child = lower.child)
+  }
+
+  private def lowerIsRedundant(upper: Aggregate, lower: Aggregate): Boolean = {
+    val upperReferencesOnlyGrouping = upper.references
+      .subsetOf(AttributeSet(lower.groupingExpressions))
+    val upperHasAggregateExpressions = upper.aggregateExpressions
+      .exists(_.find(_.isInstanceOf[AggregateExpression]).nonEmpty)
```

Comment:
This condition could be relaxed later on. It can be the same as `EliminateDistinct`, that currently allows `Max` and `Min` aggregates because their results are not changed if `distinct` is used on the inputs. Also there are some more aggregates, that can be added to `EliminateDistinct`: `BitAndAgg`, `BitOrAgg`, `CollectSet`, and perhaps some more.